lib/OpenQA/Schema/Result/Needles.pm

```diff
@@ -112,13 +112,13 @@ sub update_needle {
     my $schema = OpenQA::Schema->singleton;
     my $guard  = $schema->txn_scope_guard;
 
-    if (!-f $filename) {
-        $filename = catdir($OpenQA::Utils::sharedir, $filename);
-        if (!-f $filename) {
-            log_error(
-                "Needle file $filename not found where expected. Check $OpenQA::Utils::prjdir for distri symlinks");
-        }
-    }
+    # assume that path of the JSON file is relative to the job's needle dir or (to support legacy versions
+    # of os-autoinst) relative to the "share dir" (the $bmwqemu::vars{PRJDIR} variable in legacy os-autoinst)
+    $filename = catdir($module->job->needle_dir(), $filename) unless -f $filename;
+    $filename = catdir($OpenQA::Utils::sharedir,   $filename) unless -f $filename;
```

Comment:
we might be able to avoid this line with the additional "stat" on the file by stripping the additional directory prefixes from the legacy os-autoinst paths that match the latter parts of `needle_dir()`, WDYT?

Reply:
I'll try that.

Reply:
This involves annoying string manipulation so I've just rewrote the "stat" version so there's no additional `-f` in general and only one additional `-f` in the legacy os-autoinst case.